### PR TITLE
Calculates recommendation in a worker (closes #11):

### LINKED ISSUE #11: Run SearchRecommendation in a worker.
Rather than attempt to calculate the recommendation on a cache miss, we want to kick off a worker process to do so and immediately return a 202.

It's currently shimmed to behave similar to this, but it [calculates and caches the value in the request/response cycle before returning the 202](https://github.com/mozilla/universal-search-recommendation/commit/e2c278bf3f6430d42744c45341c710706776ba3c#diff-b5bff922b5217807d1f4e2b5bb9b0e88R110).

### PR DESCRIPTION
**Pull request by chuckharmston**
* Configures a celery container to run alongside the application container.
* Configures a redis broker for celery.
* Removes the error_on_miss option from the @memorize decorator, which is not possible to manage in a queue as originally intended due to how pickle treats differently-formed imports of the same module as separate modules.
* Updates the main route to interface directly with memcached.

r? @6a68

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
+application.config.update(
+    CELERY_BROKER_URL=conf.CELERY_BROKER_URL,
+    DEBUG=conf.DEBUG
+)
+queue = make_queue(application)
+
+
+@queue.task(name='main.recommend')
+def recommend(q, key):
+    recommendation = SearchRecommendation(q).do_search(q)
```
Comment (6a68) [inline app/main.py]: Should the worker check for the existence of the key in memcache before doing the processing work?

Comment (chuckharmston) [inline app/main.py]: I don't think so; we'll eventually want to use this to refresh cache in a [beat](http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html) process.

File: app/main.py [hunk 2]
```diff
@@ -14,14 +28,18 @@ def main():
     q = request.args.get('q')
     if not q:
         abort(400)
+    key = '_'.join([
+        conf.KEY_PREFIX,
+        hashlib.md5(str(q).encode('utf-8')).hexdigest()
```
Comment (6a68) [inline app/main.py]: Hmm, this doesn't seem quite right. Unicode input doesn't seem to be handled cleanly:

screenshot 2016-01-29 13 00 44


Traceback:
```
nginx_1     | 192.168.99.1 - - [29/Jan/2016:21:00:11 +0000] "GET /?q=%F0%9F%92%94 HTTP/1.1" 202 2 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:47.0) Gecko/20100101 Firefox/47.0" "-"
app_1       | [pid: 9|app: 0|req: 1/1] 172.17.0.5 () {42 vars in 679 bytes} [Fri Jan 29 21:00:11 2016] GET /?q=%F0%9F%92%94 => generated 2 bytes in 49 msecs (HTTP/1.0 202) 2 headers in 76 bytes (1 switches on core 0)
celery_1    | [2016-01-29 21:00:11,976: ERROR/MainProcess] Task main.recommend[ba4b1ddc-d01b-44b5-a331-52c9604f9dff] raised unexpected: IndexError('list index out of range',)
celery_1    | Traceback (most recent call last):
celery_1    |   File "/usr/local/lib/python3.5/site-packages/celery/app/trace.py", line 240, in trace_task
celery_1    |     R = retval = fun(*args, **kwargs)
celery_1    |   File "/app/util.py", line 12, in __call__
celery_1    |     return TaskBase.__call__(self, *args, **kwargs)
celery_1    |   File "/usr/local/lib/python3.5/site-packages/celery/app/trace.py", line 438, in __protected_call__
celery_1    |     return self.run(*args, **kwargs)
celery_1    |   File "/app/main.py", line 21, in recommend
celery_1    |     recommendation = SearchRecommendation(q).do_search(q)
celery_1    |   File "/app/search/recommendation.py", line 84, in do_search
celery_1    |     self.top_suggestion = self.get_top_suggestion(self.suggestions)
celery_1    |   File "/app/search/recommendation.py", line 55, in get_top_suggestion
celery_1    |     return suggestions[0]
celery_1    | IndexError: list index out of range
```

File: app/memorize.py [hunk 3]
```diff
@@ -8,14 +8,6 @@
...
 
 
-class CacheMissError(RuntimeError):
-    """
-    Exception thrown when memorized object is not found in cache and
-    error_on_miss is set to True.
-    """
-    pass
-
-
```
Comment (6a68) [inline app/memorize.py]: It seems like the docs for the `memorize` class (line 57) need to be updated as well

File: app/requirements.txt [hunk 4]
```diff
@@ -1,6 +1,7 @@
 -e git+https://github.com/linsomniac/python-memcached.git@37f55ca4ad94ca4ade30d6be28e1facb79ac3182#egg=python-memcached
```
Comment (6a68) [inline app/requirements.txt]: Not part of this PR, but I'm curious why we have to use the github python-memcached URL instead of the pip package? Is there some funky open bug?

File: app/main.py (hunk 1 above)
Comment (6a68) [inline app/main.py]: Could you explain in a bit more detail how cache invalidation is managed in this code? Maybe in the code docs?

There will be a period of time between invalidating a cached result and a fresh response being inserted into the cache. Do we do anything to avoid creating redundant jobs during that period of time? If not, the number of redundant API requests will grow with the number of users, so I guess we can ignore it for now, but might be best to be explicit about that, and file a followup bug. (More likely, there's a caching layer someplace that I'm missing :-)

**Comment by 6a68** [comment]
@chuckharmston I'm a bit unclear on when to use the `@memorize` decorator vs when to get/set values in memcached directly (as done by the worker). There's some mention of this in the commit message, but could you elaborate a little bit? Do we still need `memorize` in its current form?

**Comment by 6a68** [comment]
@chuckharmston Reassigning, assign back to me when you've answered my questions above, and I'll take another look :+1:

File: app/requirements.txt (hunk 4 above)
Comment (chuckharmston) [inline app/requirements.txt]: Yeah, we need this commit:

https://github.com/linsomniac/python-memcached/commit/94089e66cd0acca36d5206c7e62499bf12a502d6

**Comment by chuckharmston** [comment]
> @chuckharmston I'm a bit unclear on when to use the @memorize decorator vs when to get/set values in memcached directly (as done by the worker). There's some mention of this in the commit message, but could you elaborate a little bit? Do we still need memorize in its current form?

Python's `pickle` serialization library, which Celery uses to serialize tasks, is really finicky about how things are imported. If they aren't imported in the same way both before pickling and before unpickling, it won't consider them equal, and will refuse to execute.

One oddity of Pickle is that the way you import a class before you pickle one of it's instances can subtly change the pickled object. Pickle requires you to have imported the object identically both before you pickle it and before you unpickle it, else:

```
PicklingError: Can't pickle <function 'foo.Bar'>: it's not the same object as foo.Bar
```

Since `memorize` passes the wrapped function around as an object, I don't think it's actually possible to bypass this, which means that `memorize` won't work. I don't think it's a huge deal, since we're likely to want to perform optimizations in the actual view function that we may not want to do any time we make an HTTP request. We still use `memorize` to wrap external HTTP requests, [like so](https://github.com/mozilla/universal-search-recommendation/blob/cb590d54c46053a166726728893c6d26facdcc7b/app/search/suggest/bing.py#L13).

File: app/main.py (hunk 2 above)
Comment (chuckharmston) [inline app/main.py]: This actually isn't an error due to how I'm using unicode here—that's just for generating a cache key—it's that Bing's suggestion API doesn't appear to recognize/return results for it:

https://api.bing.com/osjson.aspx?query=%F0%9F%92%94

In that case, the traceback + empty response is the desired behavior.

File: app/memorize.py (hunk 3 above)
Comment (chuckharmston) [inline app/memorize.py]: Sorry, what's missing from it? The only change is that I removed `error_on_miss`, since the only place we'd want to use it is in the actual request/response cycle.

File: app/main.py (hunk 1 above)
Comment (chuckharmston) [inline app/main.py]: Right now, there is no invalidation; an infinite TTL is used. To maximize the hit rate, I'd like to track the most popular searches and have a [beat](http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html) worker invalidate and update results on a schedule, but that's orthogonal to this pull request.

Bug filed to document caching layers: https://github.com/mozilla/universal-search-recommendation/issues/27

**Comment by chuckharmston** [comment]
Back to you, @6a68
